frontends/web/src/routes/account/receive/receive.tsx

```diff
@@ -15,376 +15,309 @@
-import { apiGet } from '../../../utils/request';
-import { getScriptName, isEthereumBased } from '../utils';
-import style from './receive.module.css';
+import { PairedWarning } from './components/bb01paired';
 import { ArrowCirlceLeft, ArrowCirlceLeftActive, ArrowCirlceRight, ArrowCirlceRightActive } from '../../../components/icon';
-import { Message } from '../../../components/message/message';
+import style from './receive.module.css';
 
-interface ReceiveProps {
-    code?: string;
+interface Props {
+    code: string;
     devices: TDevices;
     accounts: accountApi.IAccount[];
     deviceIDs: string[];
 }
 
-interface State {
-    verifying: boolean;
-    activeIndex: number;
-    paired: boolean | null;
-    // index into `availableScriptTypes`, or 0 if none are available.
-    addressType: number;
-    addressDialog?: { addressType: number };
-}
-
-interface LoadedReceiveProps {
-    // first array index: address types. second array index: unused addresses of that address type.
-    receiveAddresses: accountApi.ReceiveAddressList[];
-    secureOutput: {
-        hasSecureOutput: boolean;
-        optional: boolean;
-    };
-}
-
-type Props = LoadedReceiveProps & ReceiveProps & TranslateProps;
+type AddressDialog = { addressType: number } | undefined;
 
 // For BTC/LTC: all possible address types we want to offer to the user, ordered by priority (first one is default).
 // Types that are not available in the addresses delivered by the backend should be ignored.
 const scriptTypes: accountApi.ScriptType[] = ['p2wpkh', 'p2tr', 'p2wpkh-p2sh']
 
-class Receive extends Component<Props, State> {
-    // Find index in list of receive addresses that matches the given script type, or -1 if not found.
-    private getIndexOfMatchingScriptType = (scriptType: accountApi.ScriptType): number => {
-        return this.props.receiveAddresses.findIndex(addrs => addrs.scriptType !== null && scriptType === addrs.scriptType);
-    }
-
-    // All script types that are present in the addresses delivered by the backend. Will be empty for if there are no such addresses, e.g. in Ethereum.
-    private availableScriptTypes: accountApi.ScriptType[] = scriptTypes.filter(sc => this.getIndexOfMatchingScriptType(sc) >= 0);
-    public readonly state: State = {
-        verifying: false,
-        activeIndex: 0,
-        paired: null,
-        addressType: 0,
-    }
+export const Receive: FunctionComponent<Props> = ({
+    accounts,
+    code,
+    devices,
+    deviceIDs,
+}) => {
+    const { t } = useTranslation();
+    const [verifying, setVerifying] = useState<boolean>(false);
+    const [activeIndex, setActiveIndex] = useState<number>(0);
+    // index into `availableScriptTypes`, or 0 if none are available.
+    const [addressType, setAddressType] = useState<number>(0);
+    const [addressDialog, setAddressDialog] = useState<AddressDialog>();
+    const [availableScriptTypes, setAvailableScriptTypes] = useState<accountApi.ScriptType[]>();
```

Comment:
this was not in the state before, so it shouldn't be in the state now? it is a derived value

Reply:
I agree, but `availableScriptTypes` is not just a derived value that is used for rendering but also for a dependency array of a `useEffect`, so if `availableScriptTypes` changes useEffect will execute again.

The dependency array can contain normal variables too, but then I get:
![Screen Shot 2022-03-14 at 5 05 36 PM](https://user-images.githubusercontent.com/546900/158218578-dc3473a6-1abb-43f3-869e-95d97225bfc4.png)

Changing to use `useRef`

```diff
diff --git a/frontends/web/src/routes/account/receive/receive.tsx b/frontends/web/src/routes/account/receive/receive.tsx
index a949a275..c362d86b 100644
--- a/frontends/web/src/routes/account/receive/receive.tsx
+++ b/frontends/web/src/routes/account/receive/receive.tsx
@@ -15,7 +15,7 @@
  * limitations under the License.
  */
 
-import React, { FunctionComponent, useCallback, useEffect, useState} from 'react';
+import React, { FunctionComponent, useCallback, useEffect, useRef, useState} from 'react';
 import { useTranslation } from 'react-i18next';
 import { useLoad } from '../../../hooks/api';
 import * as accountApi from '../../../api/account';
@@ -60,7 +60,6 @@ export const Receive: FunctionComponent<Props> = ({
     // index into `availableScriptTypes`, or 0 if none are available.
     const [addressType, setAddressType] = useState<number>(0);
     const [addressDialog, setAddressDialog] = useState<AddressDialog>();
-    const [availableScriptTypes, setAvailableScriptTypes] = useState<accountApi.ScriptType[]>();
     const [currentAddresses, setCurrentAddresses] = useState<accountApi.IReceiveAddress[]>();
     const [currentAddressIndex, setCurrentAddressIndex] = useState<number>(0);
 
@@ -91,16 +90,18 @@ export const Receive: FunctionComponent<Props> = ({
         return receiveAddresses.findIndex(addrs => addrs.scriptType !== null && scriptType === addrs.scriptType);
     }, [receiveAddresses]);
 
+    const availableScriptTypes = useRef<accountApi.ScriptType[]>();
+
     useEffect(() => {
         if (receiveAddresses) {
             // All script types that are present in the addresses delivered by the backend. Will be empty for if there are no such addresses, e.g. in Ethereum.
-            setAvailableScriptTypes(scriptTypes.filter(sc => getIndexOfMatchingScriptType(sc) >= 0));
+            availableScriptTypes.current = scriptTypes.filter(sc => getIndexOfMatchingScriptType(sc) >= 0);
         }
     }, [receiveAddresses, getIndexOfMatchingScriptType]);
 
     useEffect(() => {
-        if (receiveAddresses && availableScriptTypes) {
-            let addressIndex = availableScriptTypes.length > 0 ? getIndexOfMatchingScriptType(availableScriptTypes[addressType]) : 0;
+        if (receiveAddresses && availableScriptTypes.current) {
+            let addressIndex = availableScriptTypes.current.length > 0 ? getIndexOfMatchingScriptType(availableScriptTypes.current[addressType]) : 0;
             if (addressIndex === -1) {
                 addressIndex = 0;
             }
@@ -166,7 +167,7 @@ export const Receive: FunctionComponent<Props> = ({
         }
     }
 
-    const hasManyScriptTypes = availableScriptTypes && availableScriptTypes.length > 1;
+    const hasManyScriptTypes = availableScriptTypes.current && availableScriptTypes.current.length > 1;
 
     return (
         <div className="contentWithGuide">
@@ -223,7 +224,7 @@ export const Receive: FunctionComponent<Props> = ({
                                         setAddressDialog(undefined);
                                     }}>
                                         <Dialog medium title={t('receive.changeScriptType')} >
-                                            {availableScriptTypes && availableScriptTypes.map((scriptType, i) => (
+                                            {availableScriptTypes.current && availableScriptTypes.current.map((scriptType, i) => (
                                                 <div key={scriptType}>
                                                     <Radio
                                                         checked={addressDialog.addressType === i}
```

Reply:
changed in https://github.com/digitalbitbox/bitbox-wallet-app/pull/1668/commits/4cffb63e8cdd741b03e9123dcf57b616cdea6f6e

(not squashed yet)

Reply:
changed in https://github.com/digitalbitbox/bitbox-wallet-app/pull/1668/commits/4cffb63e8cdd741b03e9123dcf57b616cdea6f6e

(not squashed yet)